server/snapshot.go

```diff
@@ -326,31 +337,35 @@ func (r *raftFSM) restoreChannelsFromSnapshot(serverSnap *spb.RaftSnapshot, inNe
 		if err != nil {
 			return err
 		}
-		// Do not restore messages from snapshot if the server
-		// just started and is recovering from its own snapshot.
-		if !inNewRaftCall {
-			if err := r.restoreMsgsFromSnapshot(c, sc.First, sc.Last); err != nil {
-				return err
+		// Keep track of first/last sequence in this channel's snapshot.
+		// It can help to detect that all messages in the leader have expired
+		// and this is what we should report/use as our store first/last.
+		// This is running under the FSM lock which protects this field.
+		atomic.StoreUint64(&c.firstSeq, sc.First)
+
+		// Even on startup (when inNewRaftCall==true), we need to call
+		// restoreMsgsFromSnapshot() to make sure our store is consistent.
+		// If we skip it (like we use to), then it is possible that this
+		// node misses some messages from the snapshot and will then start
+		// to apply remaining logs. It could even then become leader in
+		// the process with missing messages. So if we need to restore some
+		// messages and fail, keep trying until a leader is avail and
+		// able to serve this node.
+		for {
+			if err := r.restoreMsgsFromSnapshot(c, sc.First, sc.Last, false); err != nil {
+				s.log.Errorf("unable to restore messages, can't start until leader is available")
+				time.Sleep(time.Second)
```

Comment:
Is a whole second required here?

Reply:
Yes.. the restore failure is either because the leader is gone, or if it is a storage failure, it is likely not going to recover anyway. If it is a failure because leader is not running/gone, a new leader election takes some time, so no reason to rush this.

Reply:
Not to say that those are the only 2 reasons, there could be NATS communication issues (like sending the request fails due to disconnect and need to wait for reconnect), etc..

Reply:
Do we ever exit this or can a server look stuck sans the log messages? Should we bail after some point?

Reply:
No, this is an infinite loop. We could exit after few attempts but the point is that server will always be stuck at this step until a leader is able to service it. I can make it exit/panic after a certain number of attempts if you think it is better.

Reply:
And we want this server to either complete this step without errors or stay there (or exit, but the step can't be skipped, so will get back there on restart).

Reply:
Any management system would be waiting for it to exit to take action IMO. And if it was a health probe it would need to be able to signal to kill the server. I think we try for say 15-3o seconds then exit(-1).